src/ex4/client/src/redux/reducers/items-entities-reducer.js

```diff
@@ -0,0 +1,65 @@
+import actionTypes from "../actions/constants";
+
+const initialState = {
+	items: [],
+	lastAction: null,
+};
+
+const itemsEntitiesReducer = (state = initialState, action) => {
+	switch (action.type) {
+		case actionTypes.LOAD_ITEMS:
+			return { ...state, items: action.payload };
+
+		case actionTypes.ADD_ITEM:
+			return {
+				...state,
+				items: [...state.items, ...action.payload],
+				lastAction: {
```

Comment:
should separate the undo functionality to a different reducer.
and why do you keep action message in the state? you can deduce it from the type in the front using simple consts, the reducer / state doesn't need to know your front or what the front is expecting to show the user - it should simply contain item data / view indicator (in different reducer like isLoading, toastr etc.)